### DISCUSSION TITLE
Possible uasyncio write improvment

### DISCUSSION THREAD

**Discussion by git-n-pissed** [micropython/micropython#10043] [Libraries & Drivers]
I have some code running on an ESP8266 which was based on an old commit of MicroPython (8f55a8fab625d645b189c471f3a32c0de7e5ef7d) which did not contain uasyncio.  I rebased my code the other day on (ac5934c96b66555ba8fe43465b34cb2b2a692ff0) and updated it to use the included uasyncio.  Unfortunately I found that picoweb kept crashing due to a memory error on the last line of `usayncio.stream.write`.  The current implementation:

```
    def write(self, buf):
        if not self.out_buf:
            # Try to write immediately to the underlying stream.
            ret = self.s.write(buf)
            if ret == len(buf):
                return
            if ret is not None:
                buf = buf[ret:]
        self.out_buf += buf
```

The problem with this is that the `=+` causes a copy of `buf` to be made even if `self.out_buf` is already empty.  This makes sense in cases where a developer desires to call `write` multiple times before `drain`, but in cases where `drain` is immediately called after `write`, the RAM used by the actual data in `buf` ends up being double what it needs to be.  This is especially evident when `awrite` / `stream_awrite` is used, because it does call `drain` immediately after calling `write`.  The following change fixed my problem and I believe prevents the described issue.

```
    def write(self, buf):
        if not self.out_buf:
            # Try to write immediately to the underlying stream.
            ret = self.s.write(buf)
            if ret == len(buf):
                return
            if ret is not None:
                buf = buf[ret:]

        if not self.out_buf:
            self.out_buf = buf
        else:
            self.out_buf += buf
```

**Comment by miketeachman** [comment]
There is an outstanding PR https://github.com/micropython/micropython/pull/7868 for this issue.
Perhaps you could add a supportive comment to this existing PR in the hope that it gets some attention.
